dashboard: auth by bearer token

blackgnezdo opened the pull request:
Just to get some early feedback...

dvyukov commented:
Overall looks good to me.

I see this still supports the old password-based auth, this is good. There are some external syzbot deployments.

Yes, I think we need caching (Auth header -> namespace). We have few QPS and this auth requires full HTTPS connection establishment. I guess in other scenarios the auth is done per-session or something. We could issue a temp session auth token, but I don't see any advantage over just sending the same Auth header + caching.
If we change any permissions we will change the config and re-deploy the app, which will wipe any in-memory caches.

That `DashboardAudience = "https://syzkaller.appspot.com/api"` does it require the app to actually be deployed on this address? If yes, then it need to be dynamic as there some other external deployments.

In the config we will need to use `OauthSubject:...` for passwords, right? And that `...` can be public, right?

dvyukov commented:
> It's just a unique token so that another token for a different audience is not used for an unrelated purpose.

ack

> Correct, the following string will be the internal GCE id assigned to the service account. It's probably visible in GCP web ui somewhere though I previously looked it up by minting and then decoding tokens.

ack

> Given the option to switch over incrementally, do you think we can start by deploying one syz-bot (e.g. OpenBSD) without the cache to measure the impact and then decide?

yes, we definitely start with switching one instance, OpenBSD is a good candidate
I am not sure how we will measure. I have a things about performance, so I will add a cache anyway. I just won't sleep well knowing that every request causes another HTTPS connection :)

blackgnezdo commented:
> yes, we definitely start with switching one instance, OpenBSD is a good candidate
> I am not sure how we will measure. I have a things about performance, so I will add a cache anyway. I just won't sleep well knowing that every request causes another HTTPS connection :)

A data point on the upper bound of savings:
```
./chk   0.03s user 0.05s system 31% cpu 0.253 total
```
That's the run time of a launching a standalone go binary running `determineAuthSubj` over a cell phone tether:
```
round-trip min/avg/max/std-dev = 11.139/82.769/191.606/58.753 ms
```

I agree that it's pointless work we can skip, just need to be careful about avoiding DoS due to cache size growth.

dvyukov commented:
> I agree that it's pointless work we can skip, just need to be careful about avoiding DoS due to cache size growth.

Something like this?
https://play.golang.org/p/LYgUAVeFBYn

blackgnezdo commented:
Question about the client side. The `key` is dragged through a bunch of places, starting with flags and configs. I'm leaning towards selecting a marker value for using in such cases. This will allow plumbing reuse without figuring out all the use cases. Then the value would be checked for this magic in `dashboard/dashapi/dashapi.go` and a metaserver queried for the auth token if need be. Do you see a better way to go about it?

dvyukov commented:
> Question about the client side. The `key` is dragged through a bunch of places, starting with flags and configs. I'm leaning towards selecting a marker value for using in such cases. This will allow plumbing reuse without figuring out all the use cases. Then the value would be checked for this magic in `dashboard/dashapi/dashapi.go` and a metaserver queried for the auth token if need be. Do you see a better way to go about it?

Doing it in dashapi.NewCustom looks reasonable.
Though it then needs to be changed to return an error?
Re the special marker value, maybe just an empty key? Otherwise we will need this special value in all configs duplicated a dozen of times.

blackgnezdo commented:
> > Question about the client side. The `key` is dragged through a bunch of places, starting with flags and configs. I'm leaning towards selecting a marker value for using in such cases. This will allow plumbing reuse without figuring out all the use cases. Then the value would be checked for this magic in `dashboard/dashapi/dashapi.go` and a metaserver queried for the auth token if need be. Do you see a better way to go about it?
> 
> Doing it in dashapi.NewCustom looks reasonable.
> Though it then needs to be changed to return an error?
> Re the special marker value, maybe just an empty key? Otherwise we will need this special value in all configs duplicated a dozen of times.

If the latest commit looks reasonable, I'll add some tests and this should be complete.

dvyukov commented on dashboard/app/auth.go:
```diff
@@ -29,31 +29,42 @@ package main
+	// TODO(blackgnezdo): cache tokens with a bit of care for concurrency.
+}
+
+func mkAuthEndpoint(u string) authEndpoint {
```
Please call this `makeAuthEndpoint`. The standard prefixes for constructors are `make` and `new`, never seen `mk`.

dvyukov commented on dashboard/app/api.go:
```diff
@@ -104,7 +104,12 @@ func handleAPI(c context.Context, r *http.Request) (reply interface{}, err error
+	auth := mkAuthEndpoint(googleTokenInfoEndpoint)
+	subj, err := auth.determineAuthSubj(r.Header["Authorization"])
+	if err != nil {
+		log.Errorf(c, "%v", err)
```
The returned error should be logged by handleJSON, so this function don't need to.

dvyukov commented on dashboard/app/auth.go:
```diff
@@ -0,0 +1,101 @@
+		log.Errorf(c, "Unexpected audience %v", claims.audience)
+		return ""
+	}
+	if claims.expiration < float64(time.Now().Unix()) {
```
Use `timeNow`. And see `advanceTime` for testing. We have time stubbed in tests.

dvyukov commented on dashboard/app/auth_test.go:
```diff
@@ -0,0 +1,91 @@
+
+import (
+	"encoding/json"
+	"github.com/google/syzkaller/dashboard/dashapi"
```
Move this to the bottom and add a blank line. We separate non-standard packages.

dvyukov commented on dashboard/app/auth.go:
```diff
@@ -57,10 +58,20 @@ func mkAuthEndpoint(u string) authEndpoint {
+// The typed representation of JWT claims.
 type jwtClaims struct {
-	Subject    string  `json:"sub"`
-	Expiration float64 `json:"exp"`
```
I don't know if  you considered this or not, but one option is to have both:
```
    Expiration float64 `json:"exp"`
    expiration time.Time
```
and make parsing function fill in the second field.

dvyukov commented on dashboard/app/auth_test.go:
```diff
@@ -1,3 +1,6 @@
+// Copyright 2017 syzkaller project authors. All rights reserved.
```
s/2017/2021/
and can be moved to the previous commit (or do you plan to squash?)

dvyukov commented on dashboard/app/auth.go:
```diff
@@ -116,34 +114,34 @@ func (auth *authEndpoint) determineAuthSubj(authHeader []string) (string, error)
 // Verifies that the given credentials are acceptable and returns the
 // corresponding namespace.
-func checkClient(name0, secretPassword, oauthSubject string) (string, error) {
+func checkClient(conf *GlobalConfig, name0, secretPassword, oauthSubject string) (string, error) {
```
We currently never pass config around and rely on the global one.
The tests use the config defined in app_test.go. So I would add there a namespace for auth testing with a set of clients with different passwords/oauth.

dvyukov commented on dashboard/dashapi/jwt.go:
```diff
@@ -0,0 +1,91 @@
+// Copyright 2017 syzkaller project authors. All rights reserved.
```
s/2017/2021/

dvyukov commented on dashboard/dashapi/jwt.go:
```diff
@@ -0,0 +1,91 @@
+			t, err := retrieveJwtToken(ctor, doer)
+			if err != nil {
+				// Can't get a new token, so returning the error preemptively.
+				return nil, err
```
This path does not unlock the mutex.

dvyukov commented on dashboard/dashapi/jwt.go:
```diff
@@ -0,0 +1,91 @@
+	if len(pieces) != 3 {
+		return time.Time{}, fmt.Errorf("unexpected number of JWT components %v", len(pieces))
+	}
+	decoded, err := base64.URLEncoding.WithPadding(base64.NoPadding).DecodeString(pieces[1])
```
Docs say there is `RawURLEncoding`:
https://golang.org/pkg/encoding/base64/#pkg-variables
Or it's not present in the appengine version?

dvyukov commented on dashboard/dashapi/jwt.go:
```diff
@@ -0,0 +1,91 @@
+	if err != nil {
+		return nil, err
+	}
+	token := string(data)
```
This can be moved few lines below.

blackgnezdo commented:
Thank you for the careful review. All but one items are addressed.

On Sun, Jul 11, 2021 at 10:51 PM Dmitry Vyukov ***@***.***>
wrote:

> ***@***.**** commented on this pull request.
> ------------------------------
>
> In dashboard/app/auth.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667625439>:
>
> >  	// Used in the config map as a prefix to distinguish auth identifiers from secret passwords
>  	// (which contain arbitrary strings, that can't have this prefix).
>  	oauthMagic = "OauthSubject:"
>  )
>
> +// Represent a verification backend.
> +type authEndpoint struct {
> +	// URL supporting tokeninfo auth2 protocol.
> +	url string
> +	// TODO(blackgnezdo): cache tokens with a bit of care for concurrency.
> +}
> +
> +func mkAuthEndpoint(u string) authEndpoint {
>
> Please call this makeAuthEndpoint. The standard prefixes for constructors
> are make and new, never seen mk.
>

Done.

------------------------------
>
> In dashboard/app/api.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667626253>:
>
> > @@ -104,7 +104,12 @@ func handleAPI(c context.Context, r *http.Request) (reply interface{}, err error
>  	client := r.PostFormValue("client")
>  	method := r.PostFormValue("method")
>  	log.Infof(c, "api %q from %q", method, client)
> -	subj := determineAuthSubj(c, r.Header["Authorization"])
> +	auth := mkAuthEndpoint(googleTokenInfoEndpoint)
> +	subj, err := auth.determineAuthSubj(r.Header["Authorization"])
> +	if err != nil {
> +		log.Errorf(c, "%v", err)
>
> The returned error should be logged by handleJSON, so this function don't
> need to.
>
Done. I was confused by a log.Errorf below (line 116)


> ------------------------------
>
> In dashboard/app/auth.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667628036>:
>
> > +		// This is a normal case when the client uses a password.
> +		return ""
> +	}
> +	// Values past this point are real authentication attempts. Whether
> +	// or not they are valid is the question.
> +	tokenValue := strings.TrimSpace(strings.TrimPrefix(authHeader[0], "Bearer"))
> +	claims, err := queryTokenInfo(tokenValue)
> +	if err != nil {
> +		log.Errorf(c, "Failed token validation %v", err)
> +		return ""
> +	}
> +	if claims.audience != dashapi.DashboardAudience {
> +		log.Errorf(c, "Unexpected audience %v", claims.audience)
> +		return ""
> +	}
> +	if claims.expiration < float64(time.Now().Unix()) {
>
> Use timeNow. And see advanceTime for testing. We have time stubbed in
> tests.
>

Hmm, this would impose the context parameter and complicate the tests with
context building. See if the way I did it is agreeable.


> ------------------------------
>
> In dashboard/app/auth_test.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667631686>:
>
> > @@ -0,0 +1,91 @@
> +package main
> +
> +import (
> +	"encoding/json"
> +	"github.com/google/syzkaller/dashboard/dashapi"
>
> Move this to the bottom and add a blank line. We separate non-standard
> packages.
>

Done.

------------------------------
>
> In dashboard/app/auth.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667632513>:
>
> >  type jwtClaims struct {
> -	Subject    string  `json:"sub"`
> -	Expiration float64 `json:"exp"`
>
> I don't know if you considered this or not, but one option is to have both:
>
>     Expiration float64 `json:"exp"`
>     expiration time.Time
>
> and make parsing function fill in the second field.
>
Interesting hack, it didn't occur to me. I'm uneasy about temporary fields
staying behind. Especially given that the "good" field is the private one.

------------------------------
>
> In dashboard/app/auth_test.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667633098>:
>
> > @@ -1,3 +1,6 @@
> +// Copyright 2017 syzkaller project authors. All rights reserved.
>
> s/2017/2021/
> and can be moved to the previous commit (or do you plan to squash?)
>

Done. I'll squash, no good reason to have this be piecemeal.

------------------------------
>
> In dashboard/app/auth.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667634276>:
>
> >  		return "", err
>  	}
>  	return oauthMagic + claims.Subject, nil
>  }
>
>  // Verifies that the given credentials are acceptable and returns the
>  // corresponding namespace.
> -func checkClient(name0, secretPassword, oauthSubject string) (string, error) {
> +func checkClient(conf *GlobalConfig, name0, secretPassword, oauthSubject string) (string, error) {
>
> We currently never pass config around and rely on the global one.
> The tests use the config defined in app_test.go. So I would add there a
> namespace for auth testing with a set of clients with different
> passwords/oauth.
>

This feels like it'd be moving away from self-contained tests? TestClient*
seem quite readable by themselves. If I moved their data out, the reader
would have to consult an external file to understand what's going on. How
strongly do you feel about this?


> ------------------------------
>
> In dashboard/dashapi/jwt.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667634718>:
>
> > @@ -0,0 +1,91 @@
> +// Copyright 2017 syzkaller project authors. All rights reserved.
>
> s/2017/2021/
>

Done.

------------------------------
>
> In dashboard/dashapi/jwt.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667639071>:
>
> > +	return &expiringToken{token, expiration}, nil
> +}
> +
> +// Augments the given doer with an authorization header carrying the
> +// given token. The token gets refreshed when it becomes stale.
> +func atachJwtToken(ctor RequestCtor, doer RequestDoer, token *expiringToken) RequestDoer {
> +	lock := sync.Mutex{}
> +	return func(req *http.Request) (*http.Response, error) {
> +		lock.Lock()
> +		if token.expiration.Before(time.Now()) {
> +			// Keeping the lock while making http request is dubious, but
> +			// making multiple concurrent requests is not any better.
> +			t, err := retrieveJwtToken(ctor, doer)
> +			if err != nil {
> +				// Can't get a new token, so returning the error preemptively.
> +				return nil, err
>
> This path does not unlock the mutex.
>

I feel really bad about the language not making such bugs impossible
without contorted extra nested func's and double checking the err.


> ------------------------------
>
> In dashboard/dashapi/jwt.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667640519>:
>
> > +	"time"
> +)
> +
> +type expiringToken struct {
> +	token      string
> +	expiration time.Time
> +}
> +
> +// Returns the unverified expiration value from the given JWT token.
> +func extractJwtExpiration(token string) (time.Time, error) {
> +	// https://datatracker.ietf.org/doc/html/rfc7519#section-3
> +	pieces := strings.Split(token, ".")
> +	if len(pieces) != 3 {
> +		return time.Time{}, fmt.Errorf("unexpected number of JWT components %v", len(pieces))
> +	}
> +	decoded, err := base64.URLEncoding.WithPadding(base64.NoPadding).DecodeString(pieces[1])
>
> Docs say there is RawURLEncoding:
> https://golang.org/pkg/encoding/base64/#pkg-variables
> Or it's not present in the appengine version?
>

Done. I missed it.

> ------------------------------
>
> In dashboard/dashapi/jwt.go
> <https://github.com/google/syzkaller/pull/2638#discussion_r667641539>:
>
> > +	const v1meta = "http://metadata.google.internal/computeMetadata/v1"
> +	req, err := ctor("GET", v1meta+"/instance/service-accounts/default/identity?audience="+DashboardAudience, nil)
> +	if err != nil {
> +		return nil, err
> +	}
> +	req.Header.Add("Metadata-Flavor", "Google")
> +	resp, err := doer(req)
> +	if err != nil {
> +		return nil, err
> +	}
> +	defer resp.Body.Close()
> +	data, err := ioutil.ReadAll(resp.Body)
> +	if err != nil {
> +		return nil, err
> +	}
> +	token := string(data)
>
> This can be moved few lines below.
>

Done. I meant to print the string instead of bytes.

-- 
nest.cx is Gmail hosted, use PGP:
https://pgp.key-server.io/0x0B1542BD8DF5A1B0
Fingerprint: 5E2B 2D0E 1E03 2046 BEC3  4D50 0B15 42BD 8DF5 A1B0
